llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp

```diff
@@ -0,0 +1,24 @@
+//=== SYCLFrameworkOptimization.cpp - Utility pass for framework optimization //
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLFrameworkOptimization.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+using namespace sycl;
+
+PreservedAnalyses AddDebugFuncAttrs::run(Module &M, ModuleAnalysisManager &) {
+  for (Function &F : M) {
+    F.addFnAttr(Attribute::NoInline);
+    F.addFnAttr(Attribute::OptimizeNone);
+  }
```

Comment:
That doesn't sound right at all. So here we are modifying **all** functions in the module - why?

There are cases when `optnone` can't be added or it will produce invalid LLVM IR:

https://github.com/intel/llvm/blob/1670049aff43dbe529a15704bb66a321696a699f/clang/lib/CodeGen/CodeGenModule.cpp#L2216-L2218

Also, as I said in another comments chain, I don't fully understand the motivation and requirement to have this pass at all and I would prefer not to have and therefore not increase maintenance cost of our compiler.

Note if we decide to keep it, there are no LIT tests for it and that doesn't seem right either.

Reply:
+1.
If we do want to preserve the attributes, we might need a way to save the attributes, remove them and then retrieve them after all optimizations have been applied. Is it a good idea to move the pipeline into sycl-post-link tool where this can be done?

Reply:
sycl-post-link question.
It isn't really a good idea to do this in sycl-post-link tool because of the following reasons:
- Some optimizations require knowledge about target's traits which are provided by [TargetInfo](https://clang.llvm.org/doxygen/classclang_1_1TargetInfo.html) class instance. We don't have a setup of TargetInfo in sycl-post-link as far as I know. So we will be obliged to add a setting up of this stuff in sycl-post-link while it is already done in Clang part.
- There is some infrastructure in clang and llvm which isn't supported in sycl-post-link. For example, you can inspect what passes are being run during compilation with `-mllvm -debug-pass=Structure` (BTW, it is broken with new pass manager, but I believe it is going to be fixed). If we move optimization passes into sycl-post-link then they become invisible for clang user and debugging becomes harder.
- Speaking from architecture point of view, it looks to me that `sycl-post-link` becomes a place where we put any SYCL stuff and it becomes overcomplicated.

About `optnone` correctness issues.
Yes, you are right. `AlwaysInlineAttr` func attr should be taken into account. Motivation is requested in corresponding JIRA tracker.

Reply:
removed this pass.